pkg/backend/sqlite/driver.go

```diff
@@ -281,82 +350,110 @@ func NewDriver(ctx context.Context, config *DriverConfig) (*Driver, error) {
+		deleteSQL:           deleteSQL,
+		createSQL:           createSQL,
+		updateSQL:           updateSQL,
+		getSizeSQL:          getSizeSQL,
+	}
+
+	if currentSchemaVersion == NewSchemaVersion(0, 2) {
+		queries.listSQL = listSqlV0_2
+		queries.countRevisionSQL = countRevisionSQLV0_2
+	}
+
 	return &Driver{
-		config: config,
+		config:               config,
+		currentSchemaVersion: currentSchemaVersion,
+		queries:              queries,
 	}, nil
 }
 
 // setup performs table setup, which may include creation of the Kine table if
 // it doesn't already exist, migrating key_value table contents to the Kine
 // table if the key_value table exists, all in a single database transaction.
 // changes are rolled back if an error occurs.
-func setup(ctx context.Context, db database.Interface) error {
+func setup(ctx context.Context, db database.Interface) (SchemaVersion, error) {
+	var currentSchemaVersion SchemaVersion
 	conn, err := db.Conn(ctx)
 	if err != nil {
-		return err
+		return currentSchemaVersion, err
 	}
 	defer conn.Close()
 
 	// Optimistically ask for the user_version without starting a transaction
-	var currentSchemaVersion SchemaVersion
-
-	row := conn.QueryRowContext(ctx, `PRAGMA user_version`)
-	if err := row.Scan(&currentSchemaVersion); err != nil {
-		return err
+	currentSchemaVersion, err = getUserVersion(conn, ctx)
+	if err != nil {
+		return currentSchemaVersion, err
 	}
 
 	if err := currentSchemaVersion.CompatibleWith(databaseSchemaVersion); err != nil {
-		return err
+		return currentSchemaVersion, err
 	}
 	if currentSchemaVersion >= databaseSchemaVersion {
-		return nil
+		return currentSchemaVersion, nil
 	}
 
 	txn, err := conn.BeginTx(ctx, nil)
 	if err != nil {
-		return err
+		return currentSchemaVersion, err
 	}
 	defer txn.Rollback()
 
-	if err := migrate(ctx, txn); err != nil {
-		return fmt.Errorf("migration failed: %w", err)
-	}
-
-	return txn.Commit()
-}
-
-// migrate tries to migrate from a version of the database
-// to the target one.
-func migrate(ctx context.Context, txn *sql.Tx) error {
-	var currentSchemaVersion SchemaVersion
-
-	row := txn.QueryRowContext(ctx, `PRAGMA user_version`)
-	if err := row.Scan(&currentSchemaVersion); err != nil {
-		return err
-	}
-
 	if err := currentSchemaVersion.CompatibleWith(databaseSchemaVersion); err != nil {
-		return err
+		if err.Error() == ErrSafeguard.Error() {
+			logrus.WithError(err).Debug("legacy node detected, not migrating")
+			return currentSchemaVersion, nil
+		}
+		return currentSchemaVersion, err
 	}
 	if currentSchemaVersion >= databaseSchemaVersion {
-		return nil
+		return currentSchemaVersion, nil
 	}
 
-	switch currentSchemaVersion {
-	case NewSchemaVersion(0, 0):
+	if currentSchemaVersion == NewSchemaVersion(0, 0) {
 		if err := applySchemaV0_1(ctx, txn); err != nil {
-			return err
+			return currentSchemaVersion, fmt.Errorf("failed to apply schema v0.1: %w", err)
+		}
+		if err := applySchemaV0_2(ctx, txn); err != nil {
```

Comment:
I think we should bump the version: while in theory compatible, given the constraint of v1.29 and v1.30 are in fact not compatible and we are taking action to make sure that the migration does not act.

The is: let's accept the fact that we made these two versions are not compatible and organize the versioning in the right semantic way to express that.